File: src/main/java/edu/uchicago/gerber/_08final/mvc/controller/Game.java
package edu.uchicago.gerber._08final.mvc.controller;

import edu.uchicago.gerber._08final.mvc.model.*;
import edu.uchicago.gerber._08final.mvc.view.GamePanel;


import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.util.LinkedList;
import java.util.Random;


// ===============================================
// == This Game class is the CONTROLLER
// ===============================================

public class Game implements Runnable, KeyListener {

    // ===============================================
    // FIELDS
    // ===============================================

    public static final Dimension DIM = new Dimension(1400,900); //the dimension of the game-screen.



    private final GamePanel gamePanel;
    //this is used throughout many classes.
    public static final Random R = new Random();

    public final static int ANIMATION_DELAY = 40; // milliseconds between frames

    public final static int FRAMES_PER_SECOND = 1000 / ANIMATION_DELAY;

    private final Thread animationThread;


    //key-codes
    private static final int
            PAUSE = 80, // p key
            QUIT = 81, // q key
            LEFT = 37, // rotate left; left arrow
            RIGHT = 39, // rotate right; right arrow
            UP = 38, // thrust; up arrow
            START = 83, // s key
            FIRE = 32, // space key
            MUTE = 77, // m-key mute

            UNIVERSE = 65, //a-key

            NUKE = 70; // f-key mute

    // for possible future use
    // HYPER = 68, 					// D key
    //ALIEN = 65;                // A key



    // ===============================================
    // ==CONSTRUCTOR
    // ===============================================

    public Game() {

        gamePanel = new GamePanel(DIM);
        gamePanel.addKeyListener(this); //Game object implements KeyListener
        //fire up the animation thread
        animationThread = new Thread(this); // pass the animation thread a runnable object, the Game object
        animationThread.setDaemon(true);
        animationThread.start();


    }

    // ===============================================
    // ==METHODS
    // ===============================================

    public static void main(String[] args) {


        //typical Swing application start; we pass EventQueue a Runnable object.
        EventQueue.invokeLater(Game::new);
    }

    // Game implements runnable, and must have run method
    @Override
    public void run() {

        // lower animation thread's priority, thereby yielding to the 'Event Dispatch Thread' or EDT
        // thread which listens to keystrokes
        animationThread.setPriority(Thread.MIN_PRIORITY);

        // and get the current time
        long startTime = System.currentTimeMillis();

        // this thread animates the scene
        while (Thread.currentThread() == animationThread) {


            //this call will cause all movables to move() and draw() themselves every ~40ms
            // see GamePanel class for details
            gamePanel.update(gamePanel.getGraphics());

            checkCollisions();
            checkNewLevel();
            checkFloaters();
            //this method will execute add() and remove() callbacks on Movable objects
            processGameOpsQueue();
            //keep track of the frame for development purposes
            CommandCenter.getInstance().incrementFrame();

            // surround the sleep() in a try/catch block
            // this simply controls delay time between
            // the frames of the animation
            try {
                // The total amount of time is guaranteed to be at least ANIMATION_DELAY long.  If processing (update)
                // between frames takes longer than ANIMATION_DELAY, then the difference between startTime -
                // System.currentTimeMillis() will be negative, then zero will be the sleep time
                startTime += ANIMATION_DELAY;

                Thread.sleep(Math.max(0,
                        startTime - System.currentTimeMillis()));
            } catch (InterruptedException e) {
                // do nothing (bury the exception), and just continue, e.g. skip this frame -- no big deal
            }
        } // end while
    } // end run

    private void checkFloaters() {

        spawnShieldFloater();
        spawnNukeFloater();
    }


    private void checkCollisions() {

        //This has order-of-growth of O(FOES * FRIENDS)
        Point pntFriendCenter, pntFoeCenter;
        int radFriend, radFoe;
        for (Movable movFriend : CommandCenter.getInstance().getMovFriends()) {
            for (Movable movFoe : CommandCenter.getInstance().getMovFoes()) {

                pntFriendCenter = movFriend.getCenter();
                pntFoeCenter = movFoe.getCenter();
                radFriend = movFriend.getRadius();
                radFoe = movFoe.getRadius();

                //detect collision
                if (pntFriendCenter.distance(pntFoeCenter) < (radFriend + radFoe)) {
                    //enqueue the friend
                    CommandCenter.getInstance().getOpsQueue().enqueue(movFriend, GameOp.Action.REMOVE);
                    //enqueue the foe
                    CommandCenter.getInstance().getOpsQueue().enqueue(movFoe, GameOp.Action.REMOVE);
                }
            }//end inner for
        }//end outer for

        //check for collisions between falcon and floaters. Order of growth of O(FLOATERS)
        Point pntFalCenter = CommandCenter.getInstance().getFalcon().getCenter();
        int radFalcon = CommandCenter.getInstance().getFalcon().getRadius();

        Point pntFloaterCenter;
        int radFloater;
        for (Movable movFloater : CommandCenter.getInstance().getMovFloaters()) {
            pntFloaterCenter = movFloater.getCenter();
            radFloater = movFloater.getRadius();
            //detect collision
            if (pntFalCenter.distance(pntFloaterCenter) < (radFalcon + radFloater)) {
                //enqueue the floater
                CommandCenter.getInstance().getOpsQueue().enqueue(movFloater, GameOp.Action.REMOVE);
            }//end if
        }//end for

    }//end meth


    //This method adds and removes movables to/from their respective linked-lists.
    private void processGameOpsQueue() {

        //deferred mutation: these operations are done AFTER we have completed our collision detection to avoid
        // mutating the movable linkedlists while iterating them above.
        while (!CommandCenter.getInstance().getOpsQueue().isEmpty()) {

            GameOp gameOp = CommandCenter.getInstance().getOpsQueue().dequeue();

            //given team, determine which linked-list this object will be added-to or removed-from
            LinkedList<Movable> list;
            Movable mov = gameOp.getMovable();
            switch (mov.getTeam()) {
                case FOE:
                    list = CommandCenter.getInstance().getMovFoes();
                    break;
                case FRIEND:
                    list = CommandCenter.getInstance().getMovFriends();
                    break;
                case FLOATER:
                    list = CommandCenter.getInstance().getMovFloaters();
                    break;
                case DEBRIS:
                default:
                    list = CommandCenter.getInstance().getMovDebris();
            }

            //pass the appropriate linked-list from above
            //this block will execute the addToGame() or removeFromGame() callbacks in the Movable models.
            GameOp.Action action = gameOp.getAction();
            if (action == GameOp.Action.ADD)
                mov.addToGame(list);
            else //REMOVE
                mov.removeFromGame(list);

        }//end while
    }





    private void spawnShieldFloater() {

        if (CommandCenter.getInstance().getFrame() % ShieldFloater.SPAWN_SHIELD_FLOATER == 0) {
            CommandCenter.getInstance().getOpsQueue().enqueue(new ShieldFloater(), GameOp.Action.ADD);
        }
    }

    private void spawnNukeFloater() {

        if (CommandCenter.getInstance().getFrame() % NukeFloater.SPAWN_NUKE_FLOATER == 0) {
            CommandCenter.getInstance().getOpsQueue().enqueue(new NukeFloater(), GameOp.Action.ADD);
        }
    }


    //this method spawns new Large (0) Asteroids
    private void spawnBigAsteroids(int num) {

        while (num-- > 0) {
            //Asteroids with size of zero are big
            CommandCenter.getInstance().getOpsQueue().enqueue(new Asteroid(0), GameOp.Action.ADD);

        }
    }





    private boolean isLevelClear() {
        //if there are no more Asteroids on the screen
        boolean asteroidFree = true;
        for (Movable movFoe : CommandCenter.getInstance().getMovFoes()) {
            if (movFoe instanceof Asteroid) {
                asteroidFree = false;
                break;
            }
        }
        return asteroidFree;
    }

    private void checkNewLevel() {

        if (isLevelClear()) {
            //currentLevel will be zero at beginning of game
            int level = CommandCenter.getInstance().getLevel();
            //award some points for having cleared the previous level
            CommandCenter.getInstance().setScore(CommandCenter.getInstance().getScore() + (10_000L * level));

            //bump the level up
            level = level + 1;
            CommandCenter.getInstance().setLevel(level);
            //spawn some big new asteroids
            spawnBigAsteroids(level);
            //make falcon invincible momentarily in case new asteroids spawn on top of him, and give player
            //time to adjust to new asteroids in game space.
            if (CommandCenter.getInstance().getFalcon().getShield() < Falcon.INITIAL_SPAWN_TIME)
                CommandCenter.getInstance().getFalcon().setShield(Falcon.INITIAL_SPAWN_TIME);
            //show "Level X" in middle of screen
            CommandCenter.getInstance().getFalcon().setShowLevel(Falcon.INITIAL_SPAWN_TIME);

        }
    }




    // ===============================================
    // KEYLISTENER METHODS
    // ===============================================

    @Override
    public void keyPressed(KeyEvent e) {
        Falcon falcon = CommandCenter.getInstance().getFalcon();
        int keyCode = e.getKeyCode();

        if (keyCode == START && CommandCenter.getInstance().isGameOver()) {
            CommandCenter.getInstance().initGame();
            return;
        }


        switch (keyCode) {
            case PAUSE:
                CommandCenter.getInstance().setPaused(!CommandCenter.getInstance().isPaused());
                break;
            case QUIT:
                System.exit(0);
                break;
            case UP:
                falcon.setThrusting(true);
                SoundLoader.playSound("whitenoise_loop.wav");
                break;
            case LEFT:
                falcon.setTurnState(Falcon.TurnState.LEFT);
                break;
            case RIGHT:
                falcon.setTurnState(Falcon.TurnState.RIGHT);
                break;


            // possible future use
            // case KILL:
            // case SHIELD:
            // case NUM_ENTER:

            default:
                break;
        }

    }

    @Override
    public void keyReleased(KeyEvent e) {
        Falcon falcon = CommandCenter.getInstance().getFalcon();
        int keyCode = e.getKeyCode();
        //show the key-code in the console
        System.out.println(keyCode);

        switch (keyCode) {
            case FIRE:
                CommandCenter.getInstance().getOpsQueue().enqueue(new Bullet(falcon), GameOp.Action.ADD);
                break;
            case NUKE:
                CommandCenter.getInstance().getOpsQueue().enqueue(new Nuke(falcon), GameOp.Action.ADD);
                break;
            //releasing either the LEFT or RIGHT arrow key will set the TurnState to IDLE
            case LEFT:
            case RIGHT:
                falcon.setTurnState(Falcon.TurnState.IDLE);
                break;
            case UP:
                falcon.setThrusting(false);
                SoundLoader.stopSound("whitenoise_loop.wav");
                break;

            case MUTE:
                CommandCenter.getInstance().setThemeMusic(!CommandCenter.getInstance().isThemeMusic());
                if (!CommandCenter.getInstance().isThemeMusic()) {
                    SoundLoader.stopSound("theme_loop.wav");
                } else {
                    SoundLoader.playSound("theme_loop.wav");
                }
                break;
            case UNIVERSE:
                CommandCenter.getInstance().cycleUniverse();
                break;
        }

    }

    @Override
    // does nothing, but we need it b/c of KeyListener contract
    public void keyTyped(KeyEvent e) {
    }

}

File: src/main/java/edu/uchicago/gerber/_08final/mvc/controller/Game.kt
package edu.uchicago.gerber._08final.mvc.controller

import edu.uchicago.gerber._08final.mvc.controller.Sound.playSound
import edu.uchicago.gerber._08final.mvc.model.*
import edu.uchicago.gerber._08final.mvc.model.Movable.Team
import edu.uchicago.gerber._08final.mvc.view.GamePanel
import java.awt.Dimension
import java.awt.EventQueue
import java.awt.Point
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.util.*
import javax.sound.sampled.Clip

fun main() {
    //typical Swing application start; we pass EventQueue a Runnable object.
    EventQueue.invokeLater(Game())
}
// ===============================================
// == This Game class is the CONTROLLER
// ===============================================
class Game : Runnable, KeyListener {
    private val gmpPanel: GamePanel
    private val animationThread: Thread

    // p key
    private val PAUSE = 80
    // q key
    private val QUIT = 81

    // rotate left; left arrow
    private val LEFT = 37
    // rotate right; right arrow
    private val RIGHT = 39
    // thrust; up arrow
    private val UP = 38

    // s key
    private val START = 83

    // space key
    private val FIRE = 32
    // m-key mute
    private val MUTE = 77

    // for possible future use
    // HYPER = 68, 					// D key
    private val ALIEN = 65 				// A key
    // SPECIAL = 70; 					// fire special weapon;  F key
    private val clpThrust: Clip
    private val clpMusicBackground: Clip

    // ===============================================
    // ==CONSTRUCTOR
    // ===============================================
    init {
        gmpPanel = GamePanel(DIM)
        gmpPanel.addKeyListener(this) //Game object implements KeyListener
        clpThrust = Sound.clipForLoopFactory("whitenoise.wav")
        clpMusicBackground = Sound.clipForLoopFactory("music-background.wav")

        //fire up the animation thread
        animationThread = Thread(this) // pass the animation thread a runnable object, the Game object
        animationThread.start()
    }




    // Game implements runnable, and must have run method
    override fun run() {

        // lower animation thread's priority, thereby yielding to the "main" aka 'Event Dispatch'
        // thread which listens to keystrokes
        animationThread.priority = Thread.MIN_PRIORITY

        // and get the current time
        var lStartTime = System.currentTimeMillis()

        // this thread animates the scene
        while (Thread.currentThread() === animationThread) {
            gmpPanel.update(gmpPanel.graphics) // see GamePanel class
            checkCollisions()
            checkNewLevel()
            checkFloaters()
            CommandCenter.incrementFrame()

            // surround the sleep() in a try/catch block
            // this simply controls delay time between
            // the frames of the animation
            try {
                // The total amount of time is guaranteed to be at least ANI_DELAY long.  If processing (update) 
                // between frames takes longer than ANI_DELAY, then the difference between lStartTime - 
                // System.currentTimeMillis() will be negative, then zero will be the sleep time
                lStartTime += ANI_DELAY.toLong()
                Thread.sleep(Math.max(0,
                        lStartTime - System.currentTimeMillis()))
            } catch (e: InterruptedException) {
                // do nothing (bury the exception), and just continue, e.g. skip this frame -- no big deal
            }
        } // end while
    } // end run

    private fun checkCollisions() {
        var pntFriendCenter: Point
        var pntFoeCenter: Point
        var radFriend: Int
        var radFoe: Int

        //This has order-of-growth of O(n^2), there is no way around this.
        for (movFriend in CommandCenter.movFriends) {
            for (movFoe in CommandCenter.movFoes) {
                pntFriendCenter = movFriend.myCenter()
                pntFoeCenter = movFoe.myCenter()
                radFriend = movFriend.myRadius()
                radFoe = movFoe.myRadius()

                //detect collision
                if (pntFriendCenter.distance(pntFoeCenter) < radFriend + radFoe) {
                    //remove the friend (so long as he is not protected)
                    if (!movFriend.isProtected()) {
                        CommandCenter.opsQueue.enqueue(movFriend, GameOp.Action.REMOVE)
                    }
                    //remove the foe
                    CommandCenter.opsQueue.enqueue(movFoe, GameOp.Action.REMOVE)
                    if (movFoe is Brick) {
                        CommandCenter.score = (CommandCenter.score + 1000)
                        playSound("rock.wav")
                    } else {
                        CommandCenter.score = (CommandCenter.score + 10)
                        playSound("kapow.wav")
                    }
                }
            } //end inner for
        } //end outer for

        //check for collisions between falcon and floaters. Order of growth of O(n) where n is number of floaters
        val pntFalCenter = CommandCenter.falcon.myCenter()
        val radFalcon = CommandCenter.falcon.myRadius()
        var pntFloaterCenter: Point
        var radFloater: Int
        for (movFloater in CommandCenter.movFloaters) {
            pntFloaterCenter = movFloater.myCenter()
            radFloater = movFloater.myRadius()

            //detect collision
            if (pntFalCenter.distance(pntFloaterCenter) < (radFalcon + radFloater)) {

                val clazz: Class<out Movable?> = movFloater.javaClass
                when (clazz.simpleName) {
                    "ShieldFloater" -> {
                        playSound("shieldup.wav")
                        CommandCenter.falcon.shield = Falcon.MAX_SHIELD
                    }

                    "NewWallFloater" -> {
                        playSound("wall.wav")
                        buildWall()
                    }
                }
                CommandCenter.opsQueue.enqueue(movFloater, GameOp.Action.REMOVE)

            } //end for
            processGameOpsQueue()
        }
    }//end meth

    private fun processGameOpsQueue() {

        //deferred mutation: these operations are done AFTER we have completed our collision detection to avoid
        // mutating the movable linkedlists while iterating them above
        while (!CommandCenter.opsQueue.isEmpty()) {
            val gameOp = CommandCenter.opsQueue.dequeue()
            val mov = gameOp?.movable
            val action = gameOp?.action
            if (mov != null) {
                when (mov.myTeam()) {
                    Team.FOE -> if (action == GameOp.Action.ADD) {
                        CommandCenter.movFoes.add(mov)
                    } else { //GameOp.Operation.REMOVE
                        CommandCenter.movFoes.remove(mov)
                        if (mov is Asteroid) spawnSmallerAsteroidsOrDebris(mov)
                    }

                    Team.FRIEND -> if (action == GameOp.Action.ADD) {
                        CommandCenter.movFriends.add(mov)
                    } else { //GameOp.Operation.REMOVE
                        if (mov is Falcon) {
                            CommandCenter.initFalconAndDecrementFalconNum()
                        } else {
                            CommandCenter.movFriends.remove(mov)
                        }
                    }

                    Team.FLOATER -> if (action == GameOp.Action.ADD) {
                        CommandCenter.movFloaters.add(mov)
                    } else { //GameOp.Operation.REMOVE
                        CommandCenter.movFloaters.remove(mov)
                    }

                    Team.DEBRIS -> if (action == GameOp.Action.ADD) {
                        CommandCenter.movDebris.add(mov)
                    } else { //GameOp.Operation.REMOVE
                        CommandCenter.movDebris.remove(mov)
                    }
                }
            }
        }
    }

    //shows how to add walls or rectangular elements one
    //brick at a time
    private fun buildWall() {
        val BRICK_SIZE = DIM.width / 30
        val ROWS = 2
        val COLS = 20
        val X_OFFSET = BRICK_SIZE * 5
        val Y_OFFSET = 50
        for (nRow in 0 until COLS) {
            for (nCol in 0 until ROWS) {
                CommandCenter.opsQueue.enqueue(
                    Brick(
                        Point(nRow * BRICK_SIZE + X_OFFSET, nCol * BRICK_SIZE + Y_OFFSET),
                        BRICK_SIZE
                    ),
                    GameOp.Action.ADD
                )
            }
        }
    }
    private fun checkFloaters() {
        spawnNewWallFloater()
        spawnShieldFloater()
    }

    private fun spawnNewWallFloater() {
        if (CommandCenter.frame % NewWallFloater.SPAWN_NEW_WALL_FLOATER == 0L && isBrickFree()) {
            CommandCenter.opsQueue.enqueue(NewWallFloater(), GameOp.Action.ADD)
        }
    }

    private fun spawnShieldFloater() {
        if (CommandCenter.frame % ShieldFloater.SPAWN_SHIELD_FLOATER == 0L) {
            CommandCenter.opsQueue.enqueue(ShieldFloater(), GameOp.Action.ADD)
        }
    }

    private fun isBrickFree(): Boolean {
        //if there are no more Bricks on the screen
        var brickFree = true
        for (movFoe in CommandCenter.movFoes) {
            if (movFoe is Brick) {
                brickFree = false
                break
            }
        }
        return brickFree
    }
    //this method spawns new Large (0) Asteroids
    private fun spawnBigAsteroids(nNum: Int) {
        var localNum = nNum
        while (localNum-- > 0) {
            //Asteroids with size of zero are big
            CommandCenter.opsQueue.enqueue(Asteroid(0), GameOp.Action.ADD)
        }
    }

    private fun spawnSmallerAsteroidsOrDebris(originalAsteroid: Asteroid) {
        var size = originalAsteroid.size
        //small asteroids
        if (size > 1) {
            CommandCenter.opsQueue.enqueue(WhiteCloudDebris(originalAsteroid), GameOp.Action.ADD)
        } else {
            //for large (0) and medium (1) sized Asteroids only, spawn 2 or 3 smaller asteroids respectively
            //We can use the existing variable (size) to do this
            size += 2
            while (size-- > 0) {
                CommandCenter.opsQueue.enqueue(Asteroid(originalAsteroid), GameOp.Action.ADD)
            }
        }
    }

    //if there are no more Asteroids on the screen
    fun isLevelClear(): Boolean {
            //if there are no more Asteroids on the screen
            var asteroidFree = true
            for (movFoe in CommandCenter.movFoes) {
                if (movFoe is Asteroid) {
                    asteroidFree = false
                    break
                }
            }
            return asteroidFree
    }

    private fun checkNewLevel() {
        if (isLevelClear()) {
            //more asteroids at each level to increase difficulty
            CommandCenter.level = CommandCenter.level + 1
            spawnBigAsteroids(CommandCenter.level)
            CommandCenter.falcon.shield = Falcon.SPAWN_INIT_VALUE
        }
    }

    // ===============================================
    // KEYLISTENER METHODS
    // ===============================================
    override fun keyPressed(e: KeyEvent) {
        val fal = CommandCenter.falcon
        val nKey = e.keyCode
        if (nKey == START && CommandCenter.isGameOver()) CommandCenter.initGame()
        when (nKey) {
            PAUSE -> {
                CommandCenter.paused = !CommandCenter.paused
                if (CommandCenter.paused) stopLoopingSounds(clpMusicBackground, clpThrust)
            }

            QUIT -> System.exit(0)
            UP -> {
                fal.thrusting = true
                if (!CommandCenter.paused && !CommandCenter.isGameOver()) clpThrust.loop(Clip.LOOP_CONTINUOUSLY)
            }

            LEFT -> fal.turnState = Falcon.TurnState.LEFT
            RIGHT -> fal.turnState = Falcon.TurnState.RIGHT

            else -> {}
        }
    }

    override fun keyReleased(e: KeyEvent) {
        val fal = CommandCenter.falcon
        val nKey = e.keyCode
        //show the key-code in the console
        //println(nKey)
        when (nKey) {
            FIRE -> {
                CommandCenter.opsQueue.enqueue(Bullet(fal), GameOp.Action.ADD)
                Sound.playSound("laser.wav")
            }

            LEFT, RIGHT -> fal.turnState = Falcon.TurnState.IDLE

            UP -> {
                fal.thrusting = false
                clpThrust.stop()
            }

            MUTE -> {
                CommandCenter.muted = !CommandCenter.muted
                if (!CommandCenter.muted) {
                    stopLoopingSounds(clpMusicBackground)
                } else {
                    clpMusicBackground.loop(Clip.LOOP_CONTINUOUSLY)
                }

            }
            else -> {}
        }
    }

    // does nothing, but we need it b/c of KeyListener contract
    override fun keyTyped(e: KeyEvent) {}

    companion object {
        // ===============================================
        // FIELDS
        // ===============================================

		val DIM = Dimension(1100, 900) //the dimension of the game.

        //this is used throughout many classes.
		val R = Random()

        const val ANI_DELAY = 40 // milliseconds between screen

        // updates (animation)
        const val FRAMES_PER_SECOND = 1000 / ANI_DELAY

        //spawn every 30 seconds
        private const val SPAWN_NEW_SHIP_FLOATER = FRAMES_PER_SECOND * 30

        // ===============================================
        // ==METHODS
        // ===============================================



        // Varargs for stopping looping-music-clips
        private fun stopLoopingSounds(vararg clpClips: Clip) {
            for (clp in clpClips) {
                clp.stop()
            }
        }
    }
}
